packages/atclient/src/atkey.c

```diff
@@ -221,9 +217,125 @@ int atclient_atkey_from_atstr(atclient_atkey *atkey, const atclient_atstr atstr)
   return atclient_atkey_from_string(atkey, atstr.str, atstr.olen);
 }
 
-int atclient_atkey_to_string(const atclient_atkey atkey, char *atkeystr, unsigned long *atkeystrlen,
-                             unsigned long *atkeystrolen) {
-  return 1; // TODO: implement
+int atclient_atkey_to_string(const atclient_atkey atkey, char *atkeystr, const unsigned long atkeystrlen, unsigned long *atkeystrolen)
+{
+    int ret = 1;
+
+    atclient_atstr string;
+    atclient_atstr_init(&string, ATKEY_GENERAL_BUFFER_SIZE);
+
+    if(atkey.metadata.iscached == 1)
+    {
+        ret = atclient_atstr_append(&string, "cached:");
+        if(ret != 0)
+        {
+            atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append_literal failed\n");
+            goto exit;
+        }
+    }
+
+    if(atkey.metadata.ispublic == 1 && atkey.atkeytype == ATCLIENT_ATKEY_TYPE_PUBLICKEY)
+    {
+        ret = atclient_atstr_append(&string, "public:");
+        if(ret != 0)
+        {
+            atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append_literal failed\n");
+            goto exit;
+        }
+    }
+    else if(atkey.metadata.ispublic == 1 || atkey.atkeytype == ATCLIENT_ATKEY_TYPE_PUBLICKEY)
+    {
+        atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_WARN, "either atkey's metadata ispublic is not set to 1 or atkey's atkeytype is not set to ATCLIENT_ATKEY_TYPE_PUBLICKEY for atkey: %.*s\n", (int) atkey.name.olen, atkey.name.str);
```

Comment:
Should we just set metadata.ispublic to 1 if the key type is ..._PUBLICKEY

Reply:
Seems like we should trust the key type in the code to be correct, over some metadata which is abstracted underneath it.

Reply:
Oh wait, I guess we wouldn't set it, but we should probably assume that the key is in fact a public key. We can warn, but still proceed with appending "public:"

Reply:
I was thinking we could have some sort of "fix atkey" function that fixes stuff like this

Reply:
> I was thinking we could have some sort of "fix atkey" function that fixes stuff like this

Yeah that would be a good parallel to the Atkey builder pattern we use in the dart sdk